conwayste/src/client.rs

```diff
@@ -1165,6 +1355,81 @@ impl MainState {
             });
         }
     }
+
+    fn handle_ui_action(&mut self, ctx: &mut Context, widget_id: WidgetID, action: UIAction) -> ConwaysteResult<()> {
+        match widget_id {
+            MAINMENU_PANE1_BUTTONYES
+            | MAINMENU_PANE1_BUTTONNO
+            | MAINMENU_TESTBUTTON  => {
+                match action {
+                    UIAction::ScreenTransition(s) => {
+                        self.screen_stack.push(s);
+                    }
+                    _ => {
+                        return Err(InvalidUIAction{reason: format!("Widget: {:?}, Action: {:?}", widget_id, action)});
+                    }
+                }
+            },
+            MAINMENU_TESTCHECKBOX => {
+                match action {
+                    UIAction::Toggle(enabled) => {
+                        self.config.modify(|settings| {
+                            settings.video.fullscreen = enabled;
+                        });
+                        self.video_settings.is_fullscreen = enabled;
+                        self.video_settings.update_fullscreen(ctx).unwrap(); // TODO: need ConwaysteError variant
+                    }
+                    _ => {
+                        return Err(InvalidUIAction{reason: format!("Widget: {:?}, Action: {:?}", widget_id, action)});
+                     }
+                }
+            },
+            MAINMENU_PANE1
+            | MAINMENU_LAYER1
+            | INGAME_LAYER1
+            | INGAME_PANE1 => {
+                return Err(NoAssociatedUIAction{
+                    reason: format!("Widget: {:?} is a Pane or Layer element and has no associated action", widget_id)
+                });
+            },
+            ui::WidgetID(_) => {},
+        }
+
+        Ok(())
+    }
+
+    fn handle_user_chat_complete(&mut self, _ctx: &mut Context) -> GameResult<()> {
+        let username = self.config.get().user.name.clone();
+        let mut msg = String::new();
+
+        if let Some(tf) = TextField::widget_from_screen_and_id(&mut self.ui_layout, Screen::Run, INGAME_PANE1_CHATBOXTEXTFIELD) {
+            if let Some(m) = tf.text() {
+                msg = format!("{}: {}", username, m);
+            }
+            tf.input_state = None;
+            tf.clear();
+        }
+
+        if !msg.is_empty() {
```

Comment:
I'd prefer returning early in this case.

```rust
if msg.is_empty() {
    return Ok(());
}
```

Reply:
Going to remove the return type since it's not useful after some of these past refactoring. Let me know if further changes are needed.